Mathlib/GroupTheory/DivisibleHull.lean

```diff
@@ -0,0 +1,301 @@
+    induction a using ind with | mk ma sa
+    induction b using ind with | mk mb sb
+    rw [mk_le_mk] at h h'
+    rw [mk_eq_mk_iff_smul_eq_smul]
+    exact le_antisymm h h'
+
+instance : LinearOrder (DivisibleHull M) where
+  le_total a b := by
+    induction a using ind with | mk ma sa
+    induction b using ind with | mk mb sb
+    exact le_total _ _
+  toDecidableLE := fun x y => by
+    -- TODO: change this
+    change Decidable (Quotient.liftOn₂' _ _ _ ?_)
+    · infer_instance
+    · intro a1 a2 b1 b2 h1 h2
+      refine lift_aux a1.1 a2.1 b1.1 b2.1 a1.2 a2.2 b1.2 b2.2 ?_ ?_
+      · rw [mk_eq_mk]
+        exact h1
+      · rw [mk_eq_mk]
+        exact h2
+
+
+theorem mk_lt_mk {m m' : M} {s s' : Submonoid.pos ℤ} : mk m s < mk m' s' ↔ s' • m < s • m' := by
+  simp_rw [lt_iff_not_ge, mk_le_mk]
+
+instance : IsOrderedAddMonoid (DivisibleHull M) where
+  add_le_add_left a b h c := by
+    induction a using ind with | mk ma sa
+    induction b using ind with | mk mb sb
+    induction c using ind with | mk mc sc
+    rw [mk_le_mk] at h
+    simp_rw [mk_add_mk, mk_le_mk, smul_add, smul_smul]
+    rw [mul_right_comm sc sa sb, add_le_add_iff_left]
+    rw [mul_right_comm sc sa sc, mul_right_comm sc sb sc]
+    rw [mul_smul _ _ ma, mul_smul _ _ mb]
+    exact zsmul_le_zsmul_right (le_of_lt (sc * sc).prop) h
+
+instance : PosSMulStrictMono ℚ (DivisibleHull M) where
+  elim a ha b c h := by
+    induction b using ind with | mk mb sb
+    induction c using ind with | mk mc sc
+    simp_rw [mk_lt_mk, Submonoid.smul_def] at h
+    simp_rw [qsmul_mk, mk_lt_mk, Submonoid.smul_def, Submonoid.mul_def, smul_smul]
+    simp_rw [mul_right_comm _ _ a.num, mul_smul _ _ mc, mul_smul _ _ mb]
+    refine (zsmul_lt_zsmul_iff_right ?_).mpr h
+    exact mul_pos (Int.natCast_pos.mpr a.den_pos) (Rat.num_pos.mpr ha)
+
+variable (M) in
+/-- The map `m ↦ m / 1` as an `OrderAddMonoidHom`. -/
+@[simps]
+def mkOrderAddMonoidHom : M →+o DivisibleHull M where
+  toFun a := mk a 1
+  map_add' a b := by simp [mk_add_mk]
+  map_zero' := by simp [mk]
+  monotone' a b h := by simpa [mk_le_mk] using h
+
+variable (M) in
+theorem mkOrderAddMonoidHom_injective : Function.Injective (mkOrderAddMonoidHom M) := by
+  intro a b
+  simp [mk_eq_mk_iff_smul_eq_smul]
+
+variable (M) in
+/-- The `OrderHom` on `ArchimedeanClass` induced by `DivisibleHull.mkOrderAddMonoidHom`. -/
+noncomputable
+def archimedeanClassOrderHom : ArchimedeanClass M →o ArchimedeanClass (DivisibleHull M) :=
+  ArchimedeanClass.orderHom (mkOrderAddMonoidHom M)
+
+theorem archimedeanClassMk_mk (m : M) (s : Submonoid.pos ℤ) :
+    ArchimedeanClass.mk (mk m s) = archimedeanClassOrderHom M (ArchimedeanClass.mk m) := by
+  trans ArchimedeanClass.mk (s.val • mk m s)
+  · rw [ArchimedeanClass.mk_smul _ (ne_of_gt s.prop)]
+  rw [archimedeanClassOrderHom, ArchimedeanClass.orderHom_mk, mkOrderAddMonoidHom_apply, zsmul_mk]
+  apply congrArg
+  simp [mk_eq_mk_iff_smul_eq_smul, Submonoid.smul_def]
+
+variable (M) in
```

Comment:
This shouldn't be an explicit variable, since it makes applying this result harder.